src/Frontend/IAST/Convert.hs

```diff
@@ -0,0 +1,89 @@
+module Frontend.IAST.Convert (convert, revert) where
+
+import qualified Frontend.IAST.Abs  as IAST
+import qualified Frontend.Zedex.Abs as ZX
+
+import Data.Complex ( Complex(..) )
+
+convert :: ZX.Program -> IAST.Program
+convert (ZX.Progr topls) = map convertTopl topls
+
+convertTopl :: ZX.Toplevel -> IAST.Toplevel
+convertTopl (ZX.ToplF name ids e) = IAST.ToplF (convertId name) (convertIds ids) (convertExpr e)
+convertTopl (ZX.ToplX c ids e)    = IAST.Topl IAST.X (complexConvert c) (convertIds ids) (convertExpr e)
+convertTopl (ZX.ToplZ c ids e)    = IAST.Topl IAST.Z (complexConvert c) (convertIds ids) (convertExpr e)
+
+convertExpr :: ZX.Expr -> IAST.Expr
+convertExpr (ZX.EVar var)      = IAST.Var (convertId var)
+convertExpr (ZX.ETup e es)     = IAST.Tup (map convertExpr (e:es))
+convertExpr ZX.EUnit           = IAST.Unit
+convertExpr (ZX.EApp e1 e2)    = IAST.App (convertExpr e1) (convertExpr e2)
+convertExpr (ZX.EComp e1 e2)   = IAST.Comp (convertExpr e1) (convertExpr e2)
+convertExpr (ZX.EHad e)        = IAST.Had (convertExpr e)
+convertExpr (ZX.EAbs ids e)    = IAST.Abs IAST.None 0 (convertIds ids) (convertExpr e)
+convertExpr (ZX.EXAbs c ids e) = IAST.Abs IAST.X (complexConvert c) (convertIds ids) (convertExpr e)
+convertExpr (ZX.EZAbs c ids e) = IAST.Abs IAST.Z (complexConvert c) (convertIds ids) (convertExpr e)
+
+complexConvert :: ZX.Complex -> Complex Double
+complexConvert (ZX.CComp (ZX.Scalar s1) (ZX.Scalar s2)) = read s1 :+ read s2
+complexConvert (ZX.CComn (ZX.Scalar s1) (ZX.Scalar s2)) = read s1 :+ negate (read s2)
+complexConvert ZX.CPi = pi :+ 0
+complexConvert ZX.CE = exp 1 :+ 0
+complexConvert (ZX.CExp c1 c2) = complexConvert c1 ** complexConvert c2
+complexConvert (ZX.CDiv c1 c2) = complexConvert c1 / complexConvert c2
+complexConvert (ZX.CMul c1 c2) = complexConvert c1 * complexConvert c2
+
+convertId :: ZX.Id -> IAST.Id
+convertId (ZX.Id ((x, y), name)) = IAST.Id (x,y) name
+
+convertIds :: [ZX.Id] -> [IAST.Id]
+convertIds = map convertId
+
+revert :: IAST.Program -> ZX.Program
+revert topls = ZX.Progr $ map revertTopl topls
+
+revertTopl :: IAST.Toplevel -> ZX.Toplevel
+revertTopl (IAST.ToplF name ids e)       = ZX.ToplF (revertId name) (revertIds ids) (revertExpr e)
+revertTopl (IAST.Topl IAST.Z c ids e)    = ZX.ToplZ (revertComplex c) (revertIds ids) (revertExpr e)
+revertTopl (IAST.Topl IAST.X c ids e)    = ZX.ToplX (revertComplex c) (revertIds ids) (revertExpr e)
+revertTopl (IAST.Topl IAST.None c ids e) = ZX.ToplF ((revertId . head) ids) (revertIds ids) (revertExpr e)
+
+revertExpr :: IAST.Expr -> ZX.Expr
+revertExpr (IAST.Var var)            = ZX.EVar (revertId var)
+revertExpr (IAST.Tup (e:es))         = ZX.ETup (revertExpr e) (map revertExpr es)
```

Comment:
This makes the function partial for tuple constructors. Do we think it is necessary to catch this? With what though..
```suggestion
revertExpr (IAST.Tup (e:es)) = ZX.ETup (revertExpr e) (map revertExpr es)
revertExpr (IAST.Tup [])     = ZX.Unit
```